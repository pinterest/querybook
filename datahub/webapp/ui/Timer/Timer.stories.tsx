import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { Timer } from './Timer';

export const _Timer = () => <Timer />;

export default {
    title: 'Stateless/Timer',
    decorators: [centered],
};
