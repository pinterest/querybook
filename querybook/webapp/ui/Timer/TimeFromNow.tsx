import moment from 'moment';
import React, { useEffect, useMemo, useState } from 'react';

import { generateFormattedDate } from 'lib/utils/datetime';

export interface TimeFromNowProps {
    /**
     * Assumed this UTC timestamp in number of seconds
     */
    timestamp: number;
    withFormattedDate?: boolean;
}

// Reference https://momentjscom.readthedocs.io/en/latest/moment/04-displaying/02-fromnow/
const FromNowTimeSegments: Array<{
    /**
     * the number of seconds where this applies
     * the array should be sorted by upper bound
     */
    upperbound: number;
    /**
     * number of seconds wait for the next update
     */
    timeToUpdate: number;
}> = [
    {
        // A few seconds
        upperbound: 44,
        timeToUpdate: 45, // 44s + 1s
    },
    {
        // A minute ago
        upperbound: 90,
        timeToUpdate: 47, // 90s - 44 + 1s
    },
    {
        // n minutes ago, up to 44 minutes
        upperbound: 2700, // 45 * 60
        timeToUpdate: 60,
    },
    {
        // an hour ago
        upperbound: 5400, // 90 * 60,
        timeToUpdate: 2701, // 5400 - 2700 + 1
    },
    {
        // n hour ago, up to 21
        upperbound: 75600, // 21 * 60 * 60
        timeToUpdate: 3600, // 60 * 60
    },
    // After this point, the update only happens once per day
];

export const TimeFromNow: React.FC<TimeFromNowProps> = ({
    timestamp,
    withFormattedDate = false,
}) => {
    const [forceRefresh, setForceRefresh] = useState(0);
    const nextTimeToUpdate = useMemo(() => {
        const nowSeconds = new Date().getTime() / 1000;
        const secondsFromNow = Math.round(nowSeconds - timestamp);

        return FromNowTimeSegments.find(
            (entry) => secondsFromNow < entry.upperbound
        )?.timeToUpdate;

        // forceRefresh is used to rerender nextTimeToUpdate
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [timestamp, forceRefresh]);

    useEffect(() => {
        let timeout: number;

        if (nextTimeToUpdate != null) {
            timeout = setTimeout(() => {
                setForceRefresh((v) => v + 1);
            }, nextTimeToUpdate * 1000);
        }

        return () => {
            if (timeout != null) {
                clearTimeout(timeout);
            }
        };
    }, [nextTimeToUpdate, forceRefresh]);

    return (
        <>
            {withFormattedDate
                ? generateFormattedDate(timestamp, 'X') + ', '
                : ''}
            {moment.utc(timestamp, 'X').fromNow()}
        </>
    );
};
