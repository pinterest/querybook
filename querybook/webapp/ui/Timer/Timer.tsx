import clsx from 'clsx';
import React, { useCallback, useImperativeHandle, useState } from 'react';

import { useInterval } from 'hooks/useInterval';

export interface ITimerProps {
    formatter?: (ts: number) => React.ReactChild;
    updater?: (ts: number) => number;

    updateFrequency?: number;
    initialValue?: number;
    className?: string;
}

const defaultFormatter = (timestamp: number) => timestamp;
const defaultUpdater = (timestamp: number) => timestamp + 1;

export interface ITimerHandles {
    updateTimer: (overrideValue: number | null) => void;
}

export const Timer = React.forwardRef<ITimerHandles, ITimerProps>(
    (
        {
            formatter = defaultFormatter,
            updater = defaultUpdater,

            updateFrequency = 1000,
            className = '',
            initialValue = 0,
        },
        ref
    ) => {
        const [value, setValue] = useState(initialValue);

        const updateTimer = useCallback(
            (overrideValue: number | null = null) => {
                setValue((oldValue) =>
                    overrideValue != null ? overrideValue : updater(oldValue)
                );
            },
            [updater]
        );
        useInterval(updateTimer, updateFrequency);

        useImperativeHandle(
            ref,
            () => ({
                updateTimer,
            }),
            [updateTimer]
        );

        const spanClassNames = clsx({
            Timer: true,
            [className]: Boolean(className),
        });
        return <span className={spanClassNames}>{formatter(value)}</span>;
    }
);
