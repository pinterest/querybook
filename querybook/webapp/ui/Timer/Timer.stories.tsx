import centered from '@storybook/addon-centered/react';
import React from 'react';

import { Timer } from './Timer';

export const _Timer = () => <Timer />;

export default {
    title: 'Stateless/Timer',
    decorators: [centered],
};
