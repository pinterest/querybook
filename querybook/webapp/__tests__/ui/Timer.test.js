import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Timer } from '../../ui/Timer/Timer';

it('renders without crashing', () => {
    shallow(<Timer />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Timer />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});
