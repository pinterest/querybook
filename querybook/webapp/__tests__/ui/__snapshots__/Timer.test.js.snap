// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<span
  className="Timer"
>
  0
</span>
`;
